plugins/src/main/java/com/google/fhir/proxy/plugin/PermissionAccessChecker.java

```diff
@@ -16,94 +16,179 @@
+    for (BundleResources bundleResources : resourcesInBundle) {
+      if (!checkUserHasRole(
+          bundleResources.getResource().fhirType(), bundleResources.getRequestType().name())) {
+        return NoOpAccessDecision.accessDenied();
+      }
     }
 
-    private boolean checkIfRoleExists(String roleName, List<String> existingRoles) {
-        if (existingRoles.contains(roleName)) {
-            return true;
-        } else {
-            return false;
-        }
+    return NoOpAccessDecision.accessGranted();
+  }
+
+  private String getRelevantRoleName(String resourceName, String methodType) {
+    return methodType + "_" + resourceName.toUpperCase();
+  }
+
+  private String getAdminRoleName(String resourceName) {
+    return "MANAGE_" + resourceName.toUpperCase();
+  }
+
+  private boolean checkIfRoleExists(String roleName, List<String> existingRoles) {
+    return existingRoles.contains(roleName);
+  }
+
+  @Named(value = "permission")
+  static class Factory implements ResourceAccessCheckerFactory {
+
+    @VisibleForTesting static final String REALM_ACCESS_CLAIM = "realm_access";
+
+    private List<String> getUserRolesFromJWT(DecodedJWT jwt) {
+      Claim claim = jwt.getClaim(REALM_ACCESS_CLAIM);
+      Map<String, Object> roles = claim.asMap();
+      List<Object> collection = roles.values().stream().collect(Collectors.toList());
+      List<String> rolesList = new ArrayList<>();
+      List<Object> collectionPart =
+          collection != null && collection.size() > 0
+              ? (List<Object>) collection.get(0)
+              : new ArrayList<>();
+      for (Object collect : collectionPart) {
+        rolesList.add(collect.toString());
+      }
+      return rolesList;
     }
 
-    @Named(value = "permission")
-    static class Factory implements AccessCheckerFactory {
-
-        private List<String> getUserRolesFromJWT(DecodedJWT jwt) {
-            Claim claim = jwt.getClaim("realm_access");
-            Map<String, Object> roles = claim.asMap();
-            List<Object> collection = roles.values()
-              .stream()
-              .collect(Collectors.toList());
-            List<String> rolesList = new ArrayList<>();
-            List<Object> collectionPart = (List<Object>) collection.get(0);
-            for (Object collect : collectionPart) {
-                rolesList.add(collect.toString());
-            }
-            return rolesList;
-        }
-
-        @Override
-        public AccessChecker create(DecodedJWT jwt, HttpFhirClient httpFhirClient, FhirContext fhirContext, PatientFinder patientFinder) {
-            List<String> userRoles = getUserRolesFromJWT(jwt);
-            return new PermissionAccessChecker(userRoles);
-        }
+    @VisibleForTesting static final String PATIENT_CLAIM = "patient_id";
+
+    private String getPatientId(DecodedJWT jwt) {
+      return FhirUtil.checkIdOrFail(JwtUtil.getClaimOrDie(jwt, PATIENT_CLAIM));
+    }
+
+    @Override
+    public AccessChecker create(
```

Comment:
How does the server code call this `create()`? [This](https://github.com/google/fhir-access-proxy/blob/d3d206e8288c895efb3b4645969112084e60c40e/server/src/main/java/com/google/fhir/proxy/BearerAuthorizationInterceptor.java#L236) is where the server calls `create()` and it is not this signature.

If you need a different helper class, you don't need to change the `AccessCheckerFactory` interface. Just create an instance of your helper class here (e.g., `ResourceFinder`, it is not expensive to build) and pass it to `PermissionAccessChecker` constructor below.

Reply:
Thanks for this design feedback, I believe it is what we were missing. Refactored to align with this